app/src/main/kotlin/compiler/intermediate_form/FunctionDependenciesAnalyzer.kt

```diff
@@ -0,0 +1,16 @@
+package compiler.intermediate_form
+
+import compiler.ast.Function
+import compiler.ast.Program
+import compiler.ast.Variable
+import compiler.common.semantic_analysis.ReferenceMap
+
+object FunctionDependenciesAnalyzer {
+    fun variablesUsedByFunctions(ast: Program): ReferenceMap<Function, Pair<List<Variable>, List<Variable>>> { // (accessed, modified)
```

Comment:
I think we shouldn't use Pair here, but some simple wrapper data class with named fields instead

Reply:
I would suggest something like `ReferenceMap<Function, ReferenceMap<Variable, VariableAccessMode>>` with `VariableAccessMode` being a two-option enum.

Reply:
Ok, then we can't have info like "modifies but doesn't read", but I think it's not necessary. ReadOnly and ReadWrite should be enough for our future usecases.